### PR TITLE
Plural: Support ordinals

### LINKED ISSUE #403: Support ordinals
Support plural type: ordinals.

**API**

```js
Globalize.pluralGenerator([ options ]);
```

Options:
- type (String) accepts "ordinal" or "cardinal" (the default).

**References**

- http://www.unicode.org/reports/tr35/tr35-numbers.html#Language_Plural_Rules

### PR DESCRIPTION
**Pull request by tniessen**
Fixes #403

### CODE REVIEW

File: doc/api/plural/plural-generator.md [hunk 1]
```diff
@@ -13,9 +13,20 @@ to return the plural group.
...
 A Number for which to return the plural group.
 
+**options** Optional
+
+A JSON object including none or any of the following options.
+
+>> **type** Optional
+>>
+>> String `cardinal` (default), or `ordinal`.
+
```
Comment (rxaviers) [inline doc/api/plural/plural-generator.md]: Please, could you move *options* above *value*? Also, replace `>>` with `>` only.

File: doc/api/plural/plural-generator.md [hunk 2]
```diff
@@ -37,6 +50,18 @@ plural( 1 );
 
 plural( 2 );
 // > "other"
+
+// Ordinals
+plural = Globalize.pluralGenerator( { type. "ordinal" } );
```
Comment (rxaviers) [inline doc/api/plural/plural-generator.md]: `type.` -> `type:`

Comment (rxaviers) [inline doc/api/plural/plural-generator.md]: The coding style you used is correct according to http://contribute.jquery.org/style-guide/js/. But, can you change `( { type: "ordinal" } )` to `({ type: "ordinal" })` please? The extra space is a new rule and it's not in use by this repo yet.

File: src/common/validate/parameter-type/plural-type.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+define([
+	"../parameter-type"
+], function( validateParameterType ) {
+
+return function( value, name ) {
+	validateParameterType(
+		value,
+		name,
+		value === undefined || value === "cardinal" || value === "ordinal",
+		"cardinal or ordinal"
```
Comment (rxaviers) [inline src/common/validate/parameter-type/plural-type.js]: Please, let's make it a little more explicit with `String "cardinal" or "ordinal"`.

File: src/plural.js [hunk 4]
```diff
@@ -41,23 +44,40 @@ Globalize.prototype.plural = function( value ) {
  * default/instance locale.
  */
 Globalize.pluralGenerator =
-Globalize.prototype.pluralGenerator = function() {
-	var cldr, plural;
+Globalize.prototype.pluralGenerator = function(options) {
```
Comment (rxaviers) [inline src/plural.js]: style, `function(options)` -> `function( options )`

File: src/plural.js [hunk 5]
```diff
@@ -41,23 +44,40 @@ Globalize.prototype.plural = function( value ) {
...
 
+	validateParameterTypePluralType( options.type, "options.type" );
+
 	validateDefaultLocale( cldr );
 
+	isOrdinal = (options.type === "ordinal");
+
 	cldr.on( "get", validateCldr );
-	cldr.supplemental( "plurals-type-cardinal/{language}" );
+	if (isOrdinal) {
```
Comment (rxaviers) [inline src/plural.js]: Please, replace `isOrdinal` with `options.type === "ordinal"`, which is pretty straightforward already.

Comment (rxaviers) [inline src/plural.js]: Nevermind, I've just noticed it's used again below.

Comment (rxaviers) [inline src/plural.js]: Style `(isOrdinal)` -> `( isOrdinal )`

**Comment by rxaviers** [comment]
Hi @tniessen, I left a couple of comments above. Great job so far.

File: doc/api/plural/plural-generator.md [hunk 6]
```diff
@@ -49,7 +74,7 @@ plural( 1 );
 // > "other"
 ```
 
-For comparison:
+For comparison (cardinals):
```
Comment (rxaviers) [inline doc/api/plural/plural-generator.md]: Let's create a similar table for ordinals?

File: README.md [hunk 7]
```diff
@@ -433,11 +433,13 @@ to you in different flavors):
...
 ### Plural module
 
-- **`.pluralGenerator()`**
+- **`.pluralGenerator( [options] )`**
 
   Return a function that returns the value's corresponding plural group: `zero`,
   `one`, `two`, `few`, `many`, or `other`.
 
+  The function may be used for cardinals or ordinals.
+
```
Comment (rxaviers) [inline README.md]: Below, we need one example using ({ type: "ordinal" }).

File: README.md [hunk 8]
```diff
@@ -448,9 +450,9 @@ to you in different flavors):
 
   [Read more...](doc/api/plural/plural-generator.md)
 
-- **`.plural( value )`**
+- **`.plural( value[, options ] )`**
 
-  Alias for `.pluralGenerator()( value )`.
+  Alias for `.pluralGenerator( options )( value )`.
```
Comment (jzaefferer) [inline README.md]: Shouldn't this also use the optional-argument notation? So:

> Alias for `.pluralGenerator( [options] )( value )`.

File: src/plural.js [hunk 9]
```diff
@@ -41,23 +44,40 @@ Globalize.prototype.plural = function( value ) {
...
+	validateParameterTypePlainObject( options, "options" );
+
+	options = options || {};
 	cldr = this.cldr;
 
+	validateParameterTypePluralType( options.type, "options.type" );
+
 	validateDefaultLocale( cldr );
 
+	isOrdinal = (options.type === "ordinal");
```
Comment (jzaefferer) [inline src/plural.js]: Unnecessary parens

File: test/functional/plural/plural.js [hunk 10]
```diff
@@ -87,6 +103,26 @@ QUnit.test( "should return plural form", function( assert ) {
...
+	assert.equal( Globalize( "ar" ).plural( 21, { type: "ordinal" } ), "other" );
+	assert.equal( Globalize( "ar" ).plural( 70, { type: "ordinal" } ), "other" );
+	assert.equal( Globalize( "ar" ).plural( 99, { type: "ordinal" } ), "other" );
+	assert.equal( Globalize( "ar" ).plural( 100, { type: "ordinal" } ), "other" );
+	assert.equal( Globalize( "ar" ).plural( 101, { type: "ordinal" } ), "other" );
+	assert.equal( Globalize( "ar" ).plural( 102, { type: "ordinal" } ), "other" );
+	assert.equal( Globalize( "ar" ).plural( 103, { type: "ordinal" } ), "other" );
+	assert.equal( Globalize( "ar" ).plural( 111, { type: "ordinal" } ), "other" );
+	assert.equal( Globalize( "ar" ).plural( 199, { type: "ordinal" } ), "other" );
+	assert.equal( Globalize( "ar" ).plural( 3.14, { type: "ordinal" } ), "other" );
```
Comment (jzaefferer) [inline test/functional/plural/plural.js]: What do these assertion accomplish when the expected value is always "other"? If this is useful at all, you might as well loop create a pluralFormatter, then loop over a number range and assert the ordinal for each.

**Comment by jzaefferer** [comment]
I'd like to see a brief definition for both cardinals and ordinals somewhere in the documentation. More examples would also help.

**Comment by tniessen** [comment]
I pushed a commit a few days ago, awaiting reviews.

File: doc/api/plural/plural-generator.md [hunk 11]
```diff
@@ -58,3 +83,13 @@ For comparison:
 | `plural( 2 )` | `other` | `few` | `two` |
 | `plural( 3 )` | `other` | `few` | `few` |
 | `plural( 5 )` | `other` | `many` | `few` |
+
+For comparison (ordinals):
+
+| | en (English) | ru (Russian) | ar (Arabic) |
+| --- | --- | --- | --- |
+| `plural( 0 )` | `other` | `other` | `other` |
```
Comment (rxaviers) [inline doc/api/plural/plural-generator.md]: All table needs the below change.
```diff
- plural( 0 )
+ plural( 0, { type: "ordinal" } )
```

File: src/plural.js [hunk 12]
```diff
@@ -41,23 +44,40 @@ Globalize.prototype.plural = function( value ) {
...
+	} else {
+		MakePlural.rules.cardinal = cldr.supplemental( "plurals-type-cardinal" );
+	}
 
-	plural = MakePlural( cldr.attributes.language, { "no_tests": true } );
+	plural = MakePlural( cldr.attributes.language, {
+		"no_tests": true,
+		"ordinals": isOrdinal,
+		"no_cardinals": isOrdinal
+	} );
```
Comment (rxaviers) [inline src/plural.js]: I feel this code needs simplification... What about the below?

```diff
  */
 Globalize.pluralGenerator =
 Globalize.prototype.pluralGenerator = function( options ) {
-	var cldr, plural, isOrdinal;
+	var cldr, isOrdinal, plural, type;
 
 	validateParameterTypePlainObject( options, "options" );
 
 	options = options || {};
+	type = options.type || "cardinal";
 	cldr = this.cldr;
 
 	validateParameterTypePluralType( options.type, "options.type" );
@@ -58,20 +59,11 @@ Globalize.prototype.pluralGenerator = function( options ) {
 
 	isOrdinal = options.type === "ordinal";
 
-	cldr.on( "get", validateCldr );
-	if ( isOrdinal ) {
-		cldr.supplemental( "plurals-type-ordinal/{language}" );
-	} else {
-		cldr.supplemental( "plurals-type-cardinal/{language}" );
-	}
-	cldr.off( "get", validateCldr );
+	cldr.once( "get", validateCldr );
+	cldr.supplemental([ "plurals-type-" + type, "{language}" ]);
 
 	MakePlural.rules = {};
-	if ( isOrdinal ) {
-		MakePlural.rules.ordinal = cldr.supplemental( "plurals-type-ordinal" );
-	} else {
-		MakePlural.rules.cardinal = cldr.supplemental( "plurals-type-cardinal" );
-	}
+	MakePlural.rules[ type ] = cldr.supplemental( "plurals-type-" + type );
 
 	plural = MakePlural( cldr.attributes.language, {
 		"no_tests": true,
```

**Comment by rxaviers** [comment]
Hi @tniessen, I added a couple of comments. Github doesn't send notifications when new commits are pushed. So, feel free to add a new comment if you have further updates. Thanks so far.

**Comment by tniessen** [comment]
@rxaviers Agree with both, new code is almost identical with your suggestion. Updated as 583ae45.

**Comment by rxaviers** [comment]
Thank you @tniessen, one last comment.

**Comment by tniessen** [comment]
@rxaviers Fixed in 7538dec, thank you for reviewing!

File: README.md (hunk 7 above)
Comment (rxaviers) [inline README.md]: :+1:

File: doc/api/plural/plural-generator.md (hunk 6 above)
Comment (rxaviers) [inline doc/api/plural/plural-generator.md]: :+1:

File: src/plural.js [hunk 13]
```diff
@@ -41,23 +44,33 @@ Globalize.prototype.plural = function( value ) {
...
-	cldr.supplemental( "plurals-type-cardinal/{language}" );
+	cldr.supplemental([ "plurals-type-" + type, "{language}" ]);
 	cldr.off( "get", validateCldr );
 
-	// Set CLDR data
-	MakePlural.rules = {
-		cardinal: cldr.supplemental( "plurals-type-cardinal" )
-	};
+	MakePlural.rules = {};
+	MakePlural.rules[type] = cldr.supplemental( "plurals-type-" + type );
```
Comment (rxaviers) [inline src/plural.js]: spaces `MakePlural.rules[ type ] =`

File: src/plural.js (hunk 12 above)
Comment (rxaviers) [inline src/plural.js]: :+1:

**Comment by tniessen** [comment]
@rxaviers My bad, sorry...  c47456f

**Comment by rxaviers** [comment]
@tniessen, no problem. Thanks landed :tada:

**Pull request closed without merging**